File: waqf-backend/app/routers/documents.py
"""
Document endpoints for Segment 2: upload (+ real OCR pipeline), queue, list,
get-by-id, and file preview streaming.

Route shapes intentionally mirror the frontend's mock API comments in
src/data/mockDocuments.ts 1:1 (GET /documents/queue, GET /documents/{id},
GET /documents, POST /documents/upload) so Segment 3/4 and the frontend
integration are a straight swap from mocks to real calls.

Validation rules (mandatory_fields_present, survey_number_format,
date_plausibility, cross_document_consistency — Segment 3's engine, see
app/services/validation.py) run right after the OCR pipeline persists
extracted fields, and again after a reviewer submits corrections since
those can change which rules pass. A document therefore leaves upload with
status="validated" (not "extracted") once at least one field was
extracted, and its validations list already populated.
"""
from __future__ import annotations

import logging
from datetime import datetime, timezone

from fastapi import APIRouter, BackgroundTasks, Depends, File, HTTPException, UploadFile, status
from sqlalchemy.orm import Session

from app.database import SessionLocal, get_db
from app.deps import get_current_user, get_current_user_flexible
from app.models import (
    DocumentStatus,
    DpdpStatus,
    ExtractedField,
    ExtractionSource,
    FieldCorrection,
    OcrSettings,
    Review,
    ReviewAction,
    ScriptType,
    User,
    WaqfDocument,
)
from app.schemas_documents import (
    DashboardStatsOut,
    DocumentDetailOut,
    ReviewOut,
    ReviewSubmitIn,
    SupportedLanguageOut,
    TranslateIn,
    TranslateOut,
    UploadDiagnostics,
    UploadResponse,
    ValidationResultOut,
    WaqfDocumentOut,
)
from app.models import ValidationResult
from app.services import dpdp, storage, translate_service, validation, vector_search
from app.services.ocr import pipeline

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/documents", tags=["documents"])

ACCEPTED_MIME_TYPES = {"image/png", "image/jpeg", "image/webp", "image/tiff", "application/pdf"}
ACCEPTED_EXTENSIONS = (".png", ".jpg", ".jpeg", ".webp", ".tif", ".tiff", ".pdf")
MAX_SIZE_BYTES = 25 * 1024 * 1024  # 25 MB — matches Upload.tsx

# A flagged document gets this many reupload attempts (on top of the
# original upload) before a reviewer is told to visit the office in person
# with the original document. Matches Upload.tsx's STARTING_REUPLOAD_ATTEMPTS.
MAX_REUPLOAD_ATTEMPTS = 3


def _is_accepted_file(filename: str, content_type: str | None) -> bool:
    if content_type in ACCEPTED_MIME_TYPES:
        return True
    return filename.lower().endswith(ACCEPTED_EXTENSIONS)


def _preview_url(document_id: str, request_token: str) -> str:
    return f"/api/v1/documents/{document_id}/file?token={request_token}"


def _to_out(document: WaqfDocument, preview_token: str | None = None) -> WaqfDocumentOut:
    out = WaqfDocumentOut.model_validate(document)
    if preview_token:
        out.preview_url = _preview_url(document.id, preview_token)
    return out


def _run_ocr_pipeline(document_id: str, raw_bytes: bytes, filename: str, mime_type: str | None) -> None:
    """Runs the (slow, multi-engine, network-bound) OCR pipeline and
    persists its results. Called as a FastAPI BackgroundTask *after* the
    upload response has already been sent, so it needs its own DB session —
    the request-scoped one from `get_db` is closed by the time this runs.

    This is the piece that used to live inline in `upload_document` before
    the response was returned. Splitting it out is what makes the upload
    endpoint itself fast: the client gets an immediate response with
    status="processing" and polls GET /documents/{id} for the real result,
    instead of holding one HTTP connection open for however long Sarvam/
    Tesseract/Gemini/Ollama take (which, per the Ollama-timeout/Gemini-503
    logs that motivated this change, could comfortably exceed any timeout
    a client reasonably sets — and did, which is what caused duplicate
    uploads: the client gave up and the user retried while the original
    request was still quietly finishing in the background).
    """
    db = SessionLocal()
    try:
        document = db.get(WaqfDocument, document_id)
        if document is None:
            logger.error("Background OCR task couldn't find document %s (deleted before it ran?)", document_id)
            return

        try:
            ocr_settings_row = db.get(OcrSettings, 1)
            fallback_threshold = (
                ocr_settings_row.ocr_fallback_threshold
                if ocr_settings_row is not None
                else pipeline.DEFAULT_OCR_FALLBACK_THRESHOLD
            )
            result = pipeline.process_document(raw_bytes, filename, mime_type, fallback_threshold)
            document.script_type = result.script_type
            document.overall_confidence = result.overall_confidence
            document.status = DocumentStatus.extracted
            document.extraction_notes = "\n".join(result.engine_notes)

            for field_name, reading in result.fields.items():
                db.add(
                    ExtractedField(
                        document_id=document.id,
                        field_name=field_name,
                        field_value=reading.value,
                        field_value_en=reading.value_en,
                        confidence=reading.confidence,
                        source=reading.source,
                    )
                )
            db.flush()  # ExtractedField rows must be visible before validation.run_validations queries them
            validation.run_validations(db, document)

            # Vector search POC — index the document, retrieve Top-K similar
            # candidates, then use the extracted Property ID as the sole
            # final verification step. Results are internal diagnostics only
            # (logged below); the API response contract is unchanged.
            fields_for_index = db.query(ExtractedField).filter(ExtractedField.document_id == document.id).all()
            vector_result = vector_search.run_similarity_search_pipeline(db, document.id, fields_for_index)
            logger.info(
                "vector_search: document %s indexed=%s candidates=%s property_status=%s matched_document=%s.",
                document.id,
                vector_result["indexed"],
                vector_result["similar_documents"],
                vector_result["property_match_status"],
                vector_result["matched_document_id"],
            )

        except Exception as exc:
            # A demo should never leave a document stuck on "processing"
            # forever because of a bad/unusual scan — land it in a flagged
            # state with zero fields rather than losing the upload.
            logger.exception("OCR pipeline failed for document %s", document.id)
            document.status = DocumentStatus.flagged
            document.overall_confidence = None
            document.extraction_notes = f"OCR pipeline raised an unhandled error: {exc}"
            # Still run the rule engine — mandatory_fields_present etc. will
            # all come back "fail" against zero extracted fields, which is
            # exactly the useful signal for a reviewer opening this
            # document manually.
            validation.run_validations(db, document)

        db.commit()
    finally:
        db.close()


@router.post("/upload", response_model=UploadResponse, status_code=status.HTTP_201_CREATED)
def upload_document(
    background_tasks: BackgroundTasks,
    file: UploadFile = File(...),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> UploadResponse:
    if not file.filename or not _is_accepted_file(file.filename, file.content_type):
        raise HTTPException(
            status.HTTP_415_UNSUPPORTED_MEDIA_TYPE,
            "Unsupported file type. Use JPG, PNG, TIFF, WEBP, or PDF.",
        )

    raw_bytes = file.file.read()
    if len(raw_bytes) > MAX_SIZE_BYTES:
        raise HTTPException(status.HTTP_413_REQUEST_ENTITY_TOO_LARGE, "File is larger than the 25 MB limit.")
    if len(raw_bytes) == 0:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "Uploaded file is empty.")

    dpdp_status, dpdp_reason = dpdp.check_dpdp_compliance(file.filename)
    document = WaqfDocument(
        filename=file.filename,
        status=DocumentStatus.processing,
        # script_type is NOT NULL; this is a provisional value, overwritten
        # once the background OCR pass detects the real one.
        script_type=ScriptType.marathi_devanagari,
        # A document counts as synthetic iff its filename matched the same
        # synthetic/sample/demo/template/test naming convention the DPDP
        # check already looks for (see app/services/dpdp.py) — reusing that
        # signal instead of hardcoding False means uploads of the Wk1-8
        # synthetic sample set (and scripts/generate_synthetic_sample.py)
        # actually get counted in reports.py's seeded_error_catch_rate,
        # which otherwise always falls back to a placeholder since nothing
        # could ever set this to True before.
        is_synthetic=dpdp_status == DpdpStatus.compliant,
        uploaded_by=current_user.email,
        mime_type=file.content_type,
        file_size_bytes=len(raw_bytes),
    )

    document.dpdp_status = dpdp_status
    document.dpdp_reason = dpdp_reason

    db.add(document)
    db.flush()  # assigns document.id without committing yet

    storage_path = storage.save_upload(document.id, file, raw_bytes)
    document.storage_path = storage_path

    # Commit now — the document must actually exist (status="processing")
    # before we return, so a client that starts polling immediately after
    # getting this response always finds something. The slow part (OCR)
    # happens after the response is sent; see _run_ocr_pipeline above.
    db.commit()
    db.refresh(document)

    background_tasks.add_task(_run_ocr_pipeline, document.id, raw_bytes, file.filename, file.content_type)

    return UploadResponse(
        document=_to_out(document, preview_token=_upload_preview_token(current_user)),
        fields=[],
        diagnostics=None,
    )


@router.post("/{document_id}/reupload", response_model=UploadResponse, status_code=status.HTTP_201_CREATED)
def reupload_document(
    document_id: str,
    background_tasks: BackgroundTasks,
    file: UploadFile = File(...),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> UploadResponse:
    """Replaces the file behind a flagged document and re-runs the OCR
    pipeline against it, capped at MAX_REUPLOAD_ATTEMPTS total attempts
    (see Dashboard.tsx's "Flagged for review" dialog). Unlike
    /documents/upload this does not create a new document row — it reuses
    document_id so the document's review history, id, and any place it's
    already linked from stay intact; only the file, extraction, and
    validation state are replaced."""
    document = db.get(WaqfDocument, document_id)
    if document is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Document not found.")

    if document.reupload_count >= MAX_REUPLOAD_ATTEMPTS:
        raise HTTPException(
            status.HTTP_409_CONFLICT,
            f"{MAX_REUPLOAD_ATTEMPTS} attempts done. Please visit the office.",
        )

    if not file.filename or not _is_accepted_file(file.filename, file.content_type):
        raise HTTPException(
            status.HTTP_415_UNSUPPORTED_MEDIA_TYPE,
            "Unsupported file type. Use JPG, PNG, TIFF, WEBP, or PDF.",
        )

    raw_bytes = file.file.read()
    if len(raw_bytes) > MAX_SIZE_BYTES:
        raise HTTPException(status.HTTP_413_REQUEST_ENTITY_TOO_LARGE, "File is larger than the 25 MB limit.")
    if len(raw_bytes) == 0:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "Uploaded file is empty.")

    dpdp_status, dpdp_reason = dpdp.check_dpdp_compliance(file.filename)

    storage_path = storage.save_upload(document.id, file, raw_bytes)

    # Clear out the previous attempt's fields/validations — they belong to
    # the old file and would otherwise sit alongside (or be confused with)
    # the new OCR pass's results.
    db.query(ExtractedField).filter(ExtractedField.document_id == document.id).delete()
    db.query(ValidationResult).filter(ValidationResult.document_id == document.id).delete()

    document.filename = file.filename
    document.status = DocumentStatus.processing
    document.storage_path = storage_path
    document.mime_type = file.content_type
    document.file_size_bytes = len(raw_bytes)
    document.dpdp_status = dpdp_status
    document.dpdp_reason = dpdp_reason
    document.overall_confidence = None
    document.extraction_notes = None
    document.reupload_count += 1

    db.commit()
    db.refresh(document)

    background_tasks.add_task(_run_ocr_pipeline, document.id, raw_bytes, file.filename, file.content_type)

    return UploadResponse(
        document=_to_out(document, preview_token=_upload_preview_token(current_user)),
        fields=[],
        diagnostics=None,
    )


def _upload_preview_token(user: User) -> str:
    """The upload response needs a bearer token embedded in previewUrl too
    (see get_current_user_flexible) — reuse the same short-lived JWT the
    client already holds isn't accessible server-side, so we mint one here
    scoped to this user, matching the one issued at login."""
    from app.security import create_access_token

    return create_access_token(subject=str(user.id), extra_claims={"role": user.role.value})


@router.get("/queue", response_model=list[WaqfDocumentOut])
def get_queue(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> list[WaqfDocumentOut]:
    """Documents still awaiting review, oldest first."""
    docs = (
        db.query(WaqfDocument)
        .filter(WaqfDocument.status.in_([DocumentStatus.extracted, DocumentStatus.validated]))
        .order_by(WaqfDocument.uploaded_at.asc())
        .all()
    )
    token = _upload_preview_token(current_user)
    return [_to_out(d, preview_token=token) for d in docs]


@router.get("", response_model=list[WaqfDocumentOut])
def list_documents(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> list[WaqfDocumentOut]:
    """Every document regardless of status, newest first. Backs the Dashboard table."""
    docs = db.query(WaqfDocument).order_by(WaqfDocument.uploaded_at.desc()).all()
    token = _upload_preview_token(current_user)
    return [_to_out(d, preview_token=token) for d in docs]


@router.get("/stats/summary", response_model=DashboardStatsOut)
def get_dashboard_stats(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> DashboardStatsOut:
    """Backs the Dashboard's four stat cards. Declared ahead of
    /{document_id} so "stats" isn't swallowed as a document id."""
    docs = db.query(WaqfDocument).all()
    pending_review = sum(1 for d in docs if d.status in (DocumentStatus.extracted, DocumentStatus.validated))
    flagged = sum(1 for d in docs if d.status == DocumentStatus.flagged)

    today = datetime.now(timezone.utc).date()
    approved_today = (
        db.query(Review)
        .filter(Review.action.in_([ReviewAction.approve, ReviewAction.correct]))
        .all()
    )
    approved_today_count = sum(1 for r in approved_today if r.reviewed_at.date() == today)

    scored = [d.overall_confidence for d in docs if d.overall_confidence is not None]
    avg_confidence = sum(scored) / len(scored) if scored else None

    return DashboardStatsOut(
        pending_review=pending_review,
        approved_today=approved_today_count,
        flagged=flagged,
        avg_confidence=avg_confidence,
    )


@router.get("/{document_id}", response_model=DocumentDetailOut)
def get_document(
    document_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> DocumentDetailOut:
    document = db.get(WaqfDocument, document_id)
    if document is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Document not found.")

    fields = db.query(ExtractedField).filter(ExtractedField.document_id == document_id).all()
    validations = db.query(ValidationResult).filter(ValidationResult.document_id == document_id).all()
    rule_order = {
        "mandatory_fields_present": 0,
        "survey_number_format": 1,
        "date_plausibility": 2,
        "cross_document_consistency": 3,
    }
    validations.sort(key=lambda v: rule_order.get(v.rule_name, 99))
    token = _upload_preview_token(current_user)
    return DocumentDetailOut(document=_to_out(document, preview_token=token), fields=fields, validations=validations)


@router.post("/{document_id}/revalidate", response_model=list[ValidationResultOut])
def revalidate_document(
    document_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> list[ValidationResultOut]:
    """Force-reruns the Segment 3 rule engine against a document's current
    ExtractedField rows and persists the result, replacing whatever's on
    record. Exists because run_validations() otherwise only fires at upload
    time or when a reviewer submits an edited field — a document whose
    ValidationResult rows are missing (e.g. processed under an earlier
    build, before validation.py was wired into the upload endpoint) has no
    other way to get them without a reviewer editing a field first. Lets
    Review.tsx offer a manual "Re-run validation" action for that case."""
    document = db.get(WaqfDocument, document_id)
    if document is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Document not found.")

    results = validation.run_validations(db, document)
    db.commit()

    rule_order = {
        "mandatory_fields_present": 0,
        "survey_number_format": 1,
        "date_plausibility": 2,
        "cross_document_consistency": 3,
    }
    results.sort(key=lambda v: rule_order.get(v.rule_name, 99))
    return results


@router.post("/{document_id}/review", response_model=ReviewOut, status_code=status.HTTP_201_CREATED)
def submit_review(
    document_id: str,
    payload: ReviewSubmitIn,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> ReviewOut:
    """Records a reviewer's approve/correct/flag decision, persists any field
    corrections as an audit trail, and moves the document out of the queue.
    Mirrors the frontend's former mock `submitReview` 1:1."""
    document = db.get(WaqfDocument, document_id)
    if document is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Document not found.")

    review = Review(
        document_id=document_id,
        reviewer_id=current_user.email,
        action=payload.action,
        notes=payload.notes,
        duration_seconds=payload.duration_seconds,
    )
    db.add(review)
    db.flush()  # assigns review.id for the FieldCorrection rows below

    if payload.corrections:
        fields_by_name = {
            f.field_name: f
            for f in db.query(ExtractedField).filter(ExtractedField.document_id == document_id).all()
        }
        for field_name, corrected_value in payload.corrections.items():
            field = fields_by_name.get(field_name)
            if field is None:
                continue
            db.add(
                FieldCorrection(
                    extracted_field_id=field.id,
                    review_id=review.id,
                    previous_value=field.field_value,
                    corrected_value=corrected_value,
                )
            )
            field.field_value = corrected_value
            field.confidence = 1.0
            field.source = ExtractionSource.reconciled

        db.flush()  # corrected field_value rows must be visible before re-validating
        validation.run_validations(db, document)

    document.status = DocumentStatus.flagged if payload.action == "flag" else DocumentStatus.reviewed

    db.commit()
    db.refresh(review)
    return ReviewOut.model_validate(review)


@router.get("/{document_id}/reviews", response_model=list[ReviewOut])
def get_document_reviews(
    document_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> list[ReviewOut]:
    """Full review history for a document, oldest first. The Dashboard uses
    this to show why a flagged document was flagged (latest action='flag')."""
    reviews = (
        db.query(Review).filter(Review.document_id == document_id).order_by(Review.reviewed_at.asc()).all()
    )
    return reviews


@router.get("/translate/languages", response_model=list[SupportedLanguageOut])
def get_translate_languages(
    current_user: User = Depends(get_current_user),
) -> list[SupportedLanguageOut]:
    """Languages the flag-reason translator (below) can translate into —
    powers the language dropdown next to a flag reason on the Dashboard."""
    return [SupportedLanguageOut(**lang) for lang in translate_service.SUPPORTED_LANGUAGES]


@router.post("/translate", response_model=TranslateOut)
def translate_flag_reason(
    payload: TranslateIn,
    current_user: User = Depends(get_current_user),
) -> TranslateOut:
    """Translates a supervisor's flag/reject reason into the language a
    reviewer picks from the Dashboard's "Flagged for review" dialog (e.g.
    Urdu -> English). Source language is auto-detected — see
    services/translate_service.py."""
    try:
        translated = translate_service.translate_text(payload.text, payload.target_language)
    except translate_service.TranslationError as exc:
        raise HTTPException(status.HTTP_502_BAD_GATEWAY, str(exc))
    return TranslateOut(translated_text=translated, target_language=payload.target_language)


@router.get("/{document_id}/file")
def get_document_file(
    document_id: str,
    current_user: User = Depends(get_current_user_flexible),
    db: Session = Depends(get_db),
):
    document = db.get(WaqfDocument, document_id)
    if document is None or not document.storage_path:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Document file not found.")

    try:
        return storage.load_file_response(document.storage_path, document.mime_type, document.filename)
    except FileNotFoundError:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Stored file is missing on disk.")

File: Frontend/waqf/src/api/documents.ts
import { apiClient } from "@/api/client";
import type {
  ExtractedField,
  FieldName,
  Review,
  ReviewAction,
  ValidationResult,
  WaqfDocument,
} from "@/types/domain";

/**
 * Real backend calls for the document pipeline. Route shapes and payloads
 * mirror what src/data/mockDocuments.ts used to simulate, so this is a
 * straight swap everywhere it's imported (see app/routers/documents.py).
 */

export interface DocumentDetail {
  document: WaqfDocument;
  fields: ExtractedField[];
  validations: ValidationResult[];
}

/** GET /documents — every document regardless of status, newest first. */
export async function getAllDocuments(): Promise<WaqfDocument[]> {
  const { data } = await apiClient.get<WaqfDocument[]>("/documents");
  return data;
}

/** GET /documents/queue — documents still awaiting review, oldest first. */
export async function getQueue(): Promise<WaqfDocument[]> {
  const { data } = await apiClient.get<WaqfDocument[]>("/documents/queue");
  return data;
}

/** GET /documents/{id} */
export async function getDocument(id: string): Promise<DocumentDetail | null> {
  try {
    const { data } = await apiClient.get<DocumentDetail>(`/documents/${id}`);
    return data;
  } catch (err) {
    const status = (err as { response?: { status?: number } }).response?.status;
    if (status === 404) return null;
    throw err;
  }
}

/** POST /documents/{id}/revalidate — force-reruns the validation-rule
 *  engine against the document's current fields, e.g. when a document has
 *  no validation results because it predates the rule engine being wired
 *  in (a reviewer editing a field would trigger this implicitly; this lets
 *  Review.tsx offer it as an explicit action instead). */
export async function revalidateDocument(documentId: string): Promise<ValidationResult[]> {
  const { data } = await apiClient.post<ValidationResult[]>(`/documents/${documentId}/revalidate`);
  return data;
}

export interface DashboardStats {
  pendingReview: number;
  approvedToday: number;
  flagged: number;
  avgConfidence: number | null;
}

/** GET /documents/stats/summary */
export async function getDashboardStats(): Promise<DashboardStats> {
  const { data } = await apiClient.get<DashboardStats>("/documents/stats/summary");
  return data;
}

export interface UploadResult {
  document: WaqfDocument;
  fields: ExtractedField[];
  diagnostics: { primaryEngine: string; notes: string[] } | null;
}

/** POST /documents/upload — multipart file upload. The backend now only
 *  saves the file and creates the document row (status="processing")
 *  before responding; the actual OCR pipeline (Sarvam Vision job polling,
 *  Tesseract/Gemini comparison, Qwen-via-Ollama field mapping, Gemini
 *  backfill/translation) runs afterward as a background task, since that
 *  chain routinely took well over two minutes and no client-side timeout
 *  could reliably outlast it. This request should come back in well under
 *  a second; the 20s timeout here is just a safety margin, not a budget
 *  for OCR. Poll the returned document's id with pollDocumentUntilReady
 *  to find out when OCR actually finishes. */
export async function uploadDocument(file: File): Promise<UploadResult> {
  const formData = new FormData();
  formData.append("file", file);
  const { data } = await apiClient.post<UploadResult>("/documents/upload", formData, {
    headers: { "Content-Type": "multipart/form-data" },
    timeout: 20_000,
  });
  return data;
}

/** POST /documents/{id}/reupload — multipart file upload that replaces the
 *  file behind a flagged document and re-runs OCR against it, reusing the
 *  same document id (unlike uploadDocument, which always creates a new
 *  document). The backend caps this at MAX_REUPLOAD_ATTEMPTS and returns a
 *  409 once exhausted; the updated document's reuploadCount tells the
 *  Dashboard's flag dialog how many attempts remain. Same "processing"/poll
 *  shape as uploadDocument — the OCR pipeline runs as a background task. */
export async function reuploadDocument(documentId: string, file: File): Promise<UploadResult> {
  const formData = new FormData();
  formData.append("file", file);
  const { data } = await apiClient.post<UploadResult>(`/documents/${documentId}/reupload`, formData, {
    headers: { "Content-Type": "multipart/form-data" },
    timeout: 20_000,
  });
  return data;
}

export interface PollOptions {
  intervalMs?: number;
  /** Give up after this long and resolve with whatever state we last saw
   *  (still "processing" is possible — the document is safe either way,
   *  just still being worked on server-side; the caller can choose to poll
   *  again later rather than treat this as failure). */
  timeoutMs?: number;
}

/** Polls GET /documents/{id} until the background OCR task moves the
 *  document out of status="processing" (into extracted/validated/flagged),
 *  or until timeoutMs elapses. Replaces the old filename+uploader+recency
 *  matching hack — polling by id is exact, so there's no ambiguity and,
 *  since the document row already exists the moment upload() returns,
 *  there's nothing for the caller to "retry" and no way to create a
 *  duplicate by polling. */
export async function pollDocumentUntilReady(
  documentId: string,
  { intervalMs = 2000, timeoutMs = 5 * 60 * 1000 }: PollOptions = {}
): Promise<DocumentDetail | null> {
  const deadline = Date.now() + timeoutMs;
  while (Date.now() < deadline) {
    const detail = await getDocument(documentId);
    if (detail && detail.document.status !== "processing") return detail;
    await new Promise((resolve) => setTimeout(resolve, intervalMs));
  }
  return null;
}

export interface SubmitReviewOptions {
  notes?: string | null;
  corrections?: Partial<Record<FieldName, string>>;
  durationSeconds?: number;
}

/** POST /documents/{id}/review */
export async function submitReview(
  documentId: string,
  action: ReviewAction,
  opts: SubmitReviewOptions = {}
): Promise<Review> {
  const { data } = await apiClient.post<Review>(`/documents/${documentId}/review`, {
    action,
    notes: opts.notes ?? null,
    corrections: opts.corrections ?? null,
    durationSeconds: opts.durationSeconds ?? null,
  });
  return data;
}

/** GET /documents/{id}/reviews — full review history, oldest first. */
export async function getDocumentReviews(documentId: string): Promise<Review[]> {
  const { data } = await apiClient.get<Review[]>(`/documents/${documentId}/reviews`);
  return data;
}

/** Latest "flag" review for a document, if any — lets the Dashboard show why
 *  a document was flagged without pulling in the whole Review workspace. */
export async function getFlagReason(
  documentId: string
): Promise<{ reason: string | null; reviewerId: string; reviewedAt: string } | null> {
  const reviews = await getDocumentReviews(documentId);
  const flagReviews = reviews.filter((r) => r.action === "flag");
  const latest = flagReviews[flagReviews.length - 1];
  if (!latest) return null;
  return { reason: latest.notes, reviewerId: latest.reviewerId, reviewedAt: latest.reviewedAt };
}

export interface SupportedLanguage {
  code: string;
  label: string;
}

/** GET /documents/translate/languages — languages the flag-reason
 *  translator can translate into, for the Dashboard's language picker. */
export async function getTranslateLanguages(): Promise<SupportedLanguage[]> {
  const { data } = await apiClient.get<SupportedLanguage[]>("/documents/translate/languages");
  return data;
}

/** POST /documents/translate — translates free text (a supervisor's flag
 *  reason) into `targetLanguage` (e.g. "en-IN", "ur-IN"). Source language
 *  is auto-detected on the backend. */
export async function translateText(text: string, targetLanguage: string): Promise<string> {
  const { data } = await apiClient.post<{ translatedText: string; targetLanguage: string }>(
    "/documents/translate",
    { text, targetLanguage }
  );
  return data.translatedText;
}
